// victory-cursor-container/lib/victory-cursor-container.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useVictoryCursorContainer = exports.VictoryCursorContainer = exports.VICTORY_CURSOR_CONTAINER_DEFAULT_PROPS = void 0;
var _react = _interopRequireDefault(require("react"));
var _victoryCore = require("victory-core");
var _defaults = _interopRequireDefault(require("lodash/defaults"));
var _isObject = _interopRequireDefault(require("lodash/isObject"));
var _cursorHelpers = require("./cursor-helpers");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const VICTORY_CURSOR_CONTAINER_DEFAULT_PROPS = exports.VICTORY_CURSOR_CONTAINER_DEFAULT_PROPS = {
  cursorLabelComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.VictoryLabel, null),
  cursorLabelOffset: {
    x: 5,
    y: -10
  },
  cursorComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.LineSegment, null)
};
const useVictoryCursorContainer = initialProps => {
  const props = {
    ...VICTORY_CURSOR_CONTAINER_DEFAULT_PROPS,
    ...initialProps
  };
  const {
    children
  } = props;
  const getCursorPosition = () => {
    const {
      cursorValue,
      defaultCursorValue,
      domain,
      cursorDimension
    } = props;
    if (cursorValue) {
      return cursorValue;
    }
    if (typeof defaultCursorValue === "number") {
      return {
        x: (domain.x[0] + domain.x[1]) / 2,
        y: (domain.y[0] + domain.y[1]) / 2,
        ...(cursorDimension ? {
          [cursorDimension]: defaultCursorValue
        } : {})
      };
    }
    return defaultCursorValue;
  };
  const getCursorLabelOffset = () => {
    const {
      cursorLabelOffset
    } = props;
    if (typeof cursorLabelOffset === "number") {
      return {
        x: cursorLabelOffset,
        y: cursorLabelOffset
      };
    }
    return cursorLabelOffset;
  };
  const getPadding = () => {
    if (props.padding === undefined) {
      const child = Array.isArray(props.children) ? props.children.find(c => {
        return (0, _isObject.default)(c.props) && c.props.padding !== undefined;
      }) : props.children;
      return _victoryCore.Helpers.getPadding(child?.props?.padding);
    }
    return _victoryCore.Helpers.getPadding(props.padding);
  };
  const getCursorElements = () => {
    const {
      scale,
      cursorLabelComponent,
      cursorLabel,
      cursorComponent,
      width,
      height,
      name,
      horizontal,
      theme
    } = props;
    const cursorDimension = _cursorHelpers.CursorHelpers.getDimension(props);
    const cursorValue = getCursorPosition();
    const cursorLabelOffset = getCursorLabelOffset();
    if (!cursorValue) {
      return [];
    }
    const newElements = [];
    const padding = getPadding();
    const cursorCoordinates = scale && "x" in scale && "y" in scale && typeof scale.y === "function" && typeof scale.x === "function" ? {
      x: horizontal ? scale.y(cursorValue.y) : scale.x(cursorValue.x),
      y: horizontal ? scale.x(cursorValue.x) : scale.y(cursorValue.y)
    } : {
      x: cursorValue.x,
      y: cursorValue.y
    };
    if (cursorLabel) {
      let labelProps = (0, _defaults.default)({
        active: true
      }, cursorLabelComponent.props, {
        x: cursorCoordinates.x + cursorLabelOffset.x,
        y: cursorCoordinates.y + cursorLabelOffset.y,
        datum: cursorValue,
        active: true,
        key: `${name}-cursor-label`
      });
      if (_victoryCore.Helpers.isTooltip(cursorLabelComponent)) {
        const tooltipTheme = theme && theme.tooltip || {};
        labelProps = (0, _defaults.default)({}, labelProps, tooltipTheme);
      }
      newElements.push( /*#__PURE__*/_react.default.cloneElement(cursorLabelComponent, (0, _defaults.default)({}, labelProps, {
        text: _victoryCore.Helpers.evaluateProp(cursorLabel, labelProps)
      })));
    }
    const cursorStyle = Object.assign({
      stroke: "black"
    }, cursorComponent.props.style);
    if (cursorDimension === "x" || cursorDimension === undefined) {
      newElements.push( /*#__PURE__*/_react.default.cloneElement(cursorComponent, {
        key: `${name}-x-cursor`,
        x1: cursorCoordinates.x,
        x2: cursorCoordinates.x,
        y1: padding.top,
        y2: (typeof height === "number" ? height : 0) - padding.bottom,
        style: cursorStyle
      }));
    }
    if (cursorDimension === "y" || cursorDimension === undefined) {
      newElements.push( /*#__PURE__*/_react.default.cloneElement(cursorComponent, {
        key: `${name}-y-cursor`,
        x1: padding.left,
        x2: (typeof width === "number" ? width : 0) - padding.right,
        y1: cursorCoordinates.y,
        y2: cursorCoordinates.y,
        style: cursorStyle
      }));
    }
    return newElements;
  };
  return {
    props,
    children: [..._react.default.Children.toArray(children), ...getCursorElements()]
  };
};
exports.useVictoryCursorContainer = useVictoryCursorContainer;
const VictoryCursorContainer = initialProps => {
  const {
    props,
    children
  } = useVictoryCursorContainer(initialProps);
  return /*#__PURE__*/_react.default.createElement(_victoryCore.VictoryContainer, props, children);
};
exports.VictoryCursorContainer = VictoryCursorContainer;
VictoryCursorContainer.role = "container";
VictoryCursorContainer.defaultEvents = initialProps => {
  const props = {
    ...VICTORY_CURSOR_CONTAINER_DEFAULT_PROPS,
    ...initialProps
  };
  const createEventHandler = (handler, disabled) =>
  // eslint-disable-next-line max-params
  (event, targetProps, eventKey, context) => disabled || props.disable ? {} : handler(event, {
    ...props,
    ...targetProps
  }, eventKey, context);
  return [{
    target: "parent",
    eventHandlers: {
      onMouseLeave: createEventHandler(_cursorHelpers.CursorHelpers.onMouseLeave),
      onMouseMove: createEventHandler(_cursorHelpers.CursorHelpers.onMouseMove),
      onTouchMove: createEventHandler(_cursorHelpers.CursorHelpers.onMouseMove)
    }
  }];
};
