// victory-brush-container/lib/victory-brush-container.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useVictoryBrushContainer = exports.VictoryBrushContainer = exports.VICTORY_BRUSH_CONTAINER_DEFAULT_PROPS = void 0;
var _react = _interopRequireDefault(require("react"));
var _victoryCore = require("victory-core");
var _brushHelpers = require("./brush-helpers");
var _defaults = _interopRequireDefault(require("lodash/defaults"));
var _reactFastCompare = _interopRequireDefault(require("react-fast-compare"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const VICTORY_BRUSH_CONTAINER_DEFAULT_PROPS = exports.VICTORY_BRUSH_CONTAINER_DEFAULT_PROPS = {
  allowDrag: true,
  allowDraw: true,
  allowResize: true,
  brushComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.Rect, null),
  brushStyle: {
    stroke: "transparent",
    fill: "black",
    fillOpacity: 0.1
  },
  handleComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.Rect, null),
  handleStyle: {
    stroke: "transparent",
    fill: "transparent"
  },
  handleWidth: 8,
  mouseMoveThreshold: 0
};
const useVictoryBrushContainer = initialProps => {
  const props = {
    ...VICTORY_BRUSH_CONTAINER_DEFAULT_PROPS,
    ...initialProps
  };
  const {
    children
  } = props;
  const getSelectBox = coordinates => {
    const {
      x,
      y
    } = coordinates;
    const {
      brushStyle,
      brushComponent,
      name
    } = props;
    const brushComponentStyle = brushComponent.props && brushComponent.props.style;
    const cursor = !props.allowDrag && !props.allowResize ? "auto" : "move";
    return x[0] !== x[1] && y[0] !== y[1] ? /*#__PURE__*/_react.default.cloneElement(brushComponent, {
      key: `${name}-brush`,
      width: Math.abs(x[1] - x[0]) || 1,
      height: Math.abs(y[1] - y[0]) || 1,
      x: Math.min(x[0], x[1]),
      y: Math.min(y[0], y[1]),
      cursor,
      style: (0, _defaults.default)({}, brushComponentStyle, brushStyle)
    }) : null;
  };
  const getCursorPointers = () => {
    const cursors = {
      yProps: "ns-resize",
      xProps: "ew-resize"
    };
    if (!props.allowResize && props.allowDrag) {
      cursors.xProps = "move";
      cursors.yProps = "move";
    } else if (!props.allowResize && !props.allowDrag) {
      cursors.xProps = "auto";
      cursors.yProps = "auto";
    }
    return cursors;
  };
  const getHandles = domain => {
    const {
      handleWidth,
      handleStyle,
      handleComponent,
      name
    } = props;
    const domainBox = _brushHelpers.BrushHelpers.getDomainBox(props, domain);
    const {
      x1,
      x2,
      y1,
      y2
    } = domainBox;
    const {
      top,
      bottom,
      left,
      right
    } = _brushHelpers.BrushHelpers.getHandles(props, domainBox);
    const width = Math.abs(x2 - x1) || 1;
    const height = Math.abs(y2 - y1) || 1;
    const handleComponentStyle = handleComponent.props && handleComponent.props.style || {};
    const style = (0, _defaults.default)({}, handleComponentStyle, handleStyle);
    const cursors = getCursorPointers();
    const yProps = {
      style,
      width,
      height: handleWidth,
      cursor: cursors.yProps
    };
    const xProps = {
      style,
      width: handleWidth,
      height,
      cursor: cursors.xProps
    };
    const handleProps = {
      top: top && Object.assign({
        x: top.x1,
        y: top.y1
      }, yProps),
      bottom: bottom && Object.assign({
        x: bottom.x1,
        y: bottom.y1
      }, yProps),
      left: left && Object.assign({
        y: left.y1,
        x: left.x1
      }, xProps),
      right: right && Object.assign({
        y: right.y1,
        x: right.x1
      }, xProps)
    };
    const handles = ["top", "bottom", "left", "right"].reduce((memo, curr) => handleProps[curr] ? memo.concat( /*#__PURE__*/_react.default.cloneElement(handleComponent, Object.assign({
      key: `${name}-handle-${curr}`
    }, handleProps[curr]))) : memo, []);
    return handles.length ? handles : null;
  };
  const getRect = () => {
    const {
      currentDomain,
      cachedBrushDomain
    } = props;
    const brushDomain = (0, _defaults.default)({}, props.brushDomain, props.domain);
    const domain = (0, _reactFastCompare.default)(brushDomain, cachedBrushDomain) ? (0, _defaults.default)({}, currentDomain, brushDomain) : brushDomain;
    const coordinates = _victoryCore.Selection.getDomainCoordinates(props, domain);
    const selectBox = getSelectBox(coordinates);
    return selectBox ? [selectBox, getHandles(domain)] : [];
  };
  return {
    props,
    children: [..._react.default.Children.toArray(children), ...getRect()]
  };
};
exports.useVictoryBrushContainer = useVictoryBrushContainer;
const VictoryBrushContainer = initialProps => {
  const {
    props,
    children
  } = useVictoryBrushContainer(initialProps);
  return /*#__PURE__*/_react.default.createElement(_victoryCore.VictoryContainer, props, children);
};
exports.VictoryBrushContainer = VictoryBrushContainer;
VictoryBrushContainer.role = "container";
VictoryBrushContainer.defaultEvents = initialProps => {
  const props = {
    ...VICTORY_BRUSH_CONTAINER_DEFAULT_PROPS,
    ...initialProps
  };
  const createEventHandler = (handler, isDisabled) =>
  // eslint-disable-next-line max-params
  (event, targetProps, eventKey, context) => props.disable || isDisabled?.(targetProps) ? {} : handler(event, {
    ...props,
    ...targetProps
  }, eventKey, context);
  return [{
    target: "parent",
    eventHandlers: {
      onMouseDown: createEventHandler(_brushHelpers.BrushHelpers.onMouseDown),
      onTouchStart: createEventHandler(_brushHelpers.BrushHelpers.onMouseDown),
      onGlobalMouseMove: createEventHandler(_brushHelpers.BrushHelpers.onGlobalMouseMove, targetProps => !targetProps.isPanning && !targetProps.isSelecting),
      onGlobalTouchMove: createEventHandler(_brushHelpers.BrushHelpers.onGlobalMouseMove, targetProps => !targetProps.isPanning && !targetProps.isSelecting),
      onGlobalMouseUp: createEventHandler(_brushHelpers.BrushHelpers.onGlobalMouseUp),
      onGlobalTouchEnd: createEventHandler(_brushHelpers.BrushHelpers.onGlobalMouseUp),
      onGlobalTouchCancel: createEventHandler(_brushHelpers.BrushHelpers.onGlobalMouseUp)
    }
  }];
};
