// @shopify/react-native-skia/lib/module/renderer/Canvas.web.js
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
import { forwardRef, useCallback, useEffect, useImperativeHandle, useMemo, useRef } from "react";
import { SkiaViewNativeId } from "../views/SkiaViewNativeId";
import { SkiaSGRoot } from "../sksg/Reconciler";
import { Skia } from "../skia";
import { SkiaPictureView } from "../views/SkiaPictureView.web";

// TODO: no need to go through the JS thread for this
const useOnSizeEvent = (resultValue, onLayout) => {
  return useCallback(event => {
    if (onLayout) {
      onLayout(event);
    }
    const {
      width,
      height
    } = event.nativeEvent.layout;
    if (resultValue) {
      resultValue.value = {
        width,
        height
      };
    }
  }, [onLayout, resultValue]);
};
export const Canvas = /*#__PURE__*/forwardRef(({
  mode,
  debug,
  opaque,
  children,
  onSize,
  onLayout: _onLayout,
  ...viewProps
}, ref) => {
  const viewRef = useRef(null);
  const rafId = useRef(null);
  const onLayout = useOnSizeEvent(onSize, _onLayout);
  // Native ID
  const nativeId = useMemo(() => {
    return SkiaViewNativeId.current++;
  }, []);

  // Root
  const root = useMemo(() => new SkiaSGRoot(Skia), []);

  // Render effects
  useEffect(() => {
    root.render(children);
    if (viewRef.current) {
      viewRef.current.setPicture(root.getPicture());
    }
  }, [children, root]);
  useEffect(() => {
    return () => {
      root.unmount();
    };
  }, [root]);
  const requestRedraw = useCallback(() => {
    rafId.current = requestAnimationFrame(() => {
      root.render(children);
      if (viewRef.current) {
        viewRef.current.setPicture(root.getPicture());
      }
      if (mode === "continuous") {
        requestRedraw();
      }
    });
  }, [children, mode, root]);
  useEffect(() => {
    if (mode === "continuous") {
      requestRedraw();
    }
    return () => {
      if (rafId.current !== null) {
        cancelAnimationFrame(rafId.current);
      }
    };
  }, [mode, requestRedraw]);

  // Component methods
  useImperativeHandle(ref, () => ({
    makeImageSnapshot: rect => {
      return SkiaViewApi.makeImageSnapshot(nativeId, rect);
    },
    makeImageSnapshotAsync: rect => {
      return SkiaViewApi.makeImageSnapshotAsync(nativeId, rect);
    },
    redraw: () => {
      var _viewRef$current;
      (_viewRef$current = viewRef.current) === null || _viewRef$current === void 0 || _viewRef$current.redraw();
    },
    getNativeId: () => {
      return nativeId;
    }
  }));
  return /*#__PURE__*/React.createElement(SkiaPictureView, _extends({
    ref: viewRef,
    collapsable: false,
    nativeID: `${nativeId}`,
    debug: debug,
    opaque: opaque,
    onLayout: onLayout
  }, viewProps));
});
