// @shopify/react-native-skia/lib/module/skia/types/Image/ColorType.js
export let ColorType = /*#__PURE__*/function (ColorType) {
  ColorType[ColorType["Unknown"] = 0] = "Unknown";
  ColorType[ColorType["Alpha_8"] = 1] = "Alpha_8";
  ColorType[ColorType["RGB_565"] = 2] = "RGB_565";
  ColorType[ColorType["ARGB_4444"] = 3] = "ARGB_4444";
  ColorType[ColorType["RGBA_8888"] = 4] = "RGBA_8888";
  ColorType[ColorType["RGB_888x"] = 5] = "RGB_888x";
  ColorType[ColorType["BGRA_8888"] = 6] = "BGRA_8888";
  ColorType[ColorType["RGBA_1010102"] = 7] = "RGBA_1010102";
  ColorType[ColorType["BGRA_1010102"] = 8] = "BGRA_1010102";
  ColorType[ColorType["RGB_101010x"] = 9] = "RGB_101010x";
  ColorType[ColorType["BGR_101010x"] = 10] = "BGR_101010x";
  ColorType[ColorType["BGR_101010x_XR"] = 11] = "BGR_101010x_XR";
  ColorType[ColorType["BGRA_10101010_XR"] = 12] = "BGRA_10101010_XR";
  ColorType[ColorType["RGBA_10x6"] = 13] = "RGBA_10x6";
  ColorType[ColorType["Gray_8"] = 14] = "Gray_8";
  ColorType[ColorType["RGBA_F16Norm"] = 15] = "RGBA_F16Norm";
  ColorType[ColorType["RGBA_F16"] = 16] = "RGBA_F16";
  ColorType[ColorType["RGB_F16F16F16x"] = 17] = "RGB_F16F16F16x";
  ColorType[ColorType["RGBA_F32"] = 18] = "RGBA_F32";
  return ColorType;
}({}); // pixel using C float for red, green, blue, alpha; in 128-bit word
